### PR TITLE
[#161] Pushdown equivalence on predicate for outer/semi join

### PR DESCRIPTION
**Pull request by Seaven**

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java [hunk 1]
```diff
@@ -49,4 +49,37 @@ public boolean check(final OptExpression input, OptimizerContext context) {
         }
         return Lists.newArrayList(root);
     }
+
+    ScalarOperator equivalenceDeriveOnPredicate(ScalarOperator on, OptExpression joinOpt, LogicalJoinOperator join) {
+        // For SQl: select * from t1 left join t2 on t1.id = t2.id where t1.id > 1
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java:54]: Update comment

File: fe/fe-core/src/test/resources/sql/optimized-plan/predicate-pushdown.sql [hunk 2]
```diff
@@ -130,7 +130,7 @@ select v1 from t0 left join t1 on v1 = v4 and v1 = 1
 LEFT OUTER JOIN (join-predicate [1: v1 = 4: v4 AND 1: v1 = 1] post-join-predicate [null])
     SCAN (columns[1: v1] predicate[null])
     EXCHANGE SHUFFLE[4]
-        SCAN (columns[4: v4] predicate[null])
+        SCAN (columns[4: v4] predicate[4: v4 = 1])
```
Comment (kangkaisen) [inline fe/fe-core/src/test/resources/sql/optimized-plan/predicate-pushdown.sql:133]: `PlanFragmentTest.testJoinPredicateTransitivity` test need update

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java [hunk 3]
```diff
@@ -49,4 +49,37 @@ public boolean check(final OptExpression input, OptimizerContext context) {
         }
         return Lists.newArrayList(root);
     }
+
+    ScalarOperator equivalenceDeriveOnPredicate(ScalarOperator on, OptExpression joinOpt, LogicalJoinOperator join) {
+        // For SQl: select * from t1 left join t2 on t1.id = t2.id and t1.id > 1
+        // Infer t2.id > 1 and Push down it to right child
+        if (!join.getJoinType().isInnerJoin() && !join.getJoinType().isSemiJoin() &&
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java:56]: isCrossJoin is enough?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java:56]: Anti join can't

**Review by kangkaisen — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java (hunk 3 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java:56]: isSemiJoin() constans antiJoins now ~

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java:56]: It's bug! I will modify the method

**Review by Youngwb — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**